File: backend/src/main/java/com/tabletennis/app/models/UniqueGame.java
package com.tabletennis.app.models;

import java.time.LocalDateTime;
import java.util.Set;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;

@Entity
@Table(name = "uniqueGames")
public class UniqueGame {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    private int highestRound;

    @NotNull
    private boolean isFinished;

    @NotNull
    private Long userId;

    @NotNull
    private LocalDateTime startTime;

    @ManyToMany(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
    @JoinTable(
        name = "uniqueGame_players", 
        joinColumns = @JoinColumn(name = "uniqueGame_id"), 
        inverseJoinColumns = @JoinColumn(name = "player_id")
    )
    private Set<Player> players;

    public void setId(Long id) {
        this.id = id;
    }
    public Long getId() {
        return id;
    }

    public int getHighestRound() {
        return highestRound;
    }
    public void setHighestRound(int highestRound) {
        this.highestRound = highestRound;
    }

    public boolean isFinished() {
        return isFinished;
    }
    public void setFinished(boolean isFinished) {
        this.isFinished = isFinished;
    }

    public Long isUserId() {return userId;}
    public void setUserId(Long userId) {this.userId = userId;}

    public LocalDateTime getStartTime() {
        return startTime;
    }

    public void setStartTime(LocalDateTime startTime) {
        this.startTime = startTime;
    }

    public Set<Player> getPlayers() {
        return players;
    }

    public void setPlayers(Set<Player> players) {
        this.players = players;
    }
}

File: mobile_application/lib/entities/uniqueGame.dart
class UniqueGame {
  int _id;
  int _highestRound;
  int _userId;
  bool _isFinished;
  DateTime _startTime;
  List<int> _players; 
  int _minigameId;

  UniqueGame(this._id, this._highestRound, this._userId, this._isFinished, this._startTime, this._players, this._minigameId,);

  int get id => _id;
  set id(int value) => _id = value;

  int get highestRound => _highestRound;
  set highestRound(int value) => _highestRound = value;

  int get userId => _userId;
  set userId(int value) => _userId = value;

  bool get isFinished => _isFinished;
  set isFinished(bool value) => _isFinished = value;

  DateTime get startTime => _startTime;
  set startTime(DateTime value) => _startTime = value;

  List<int> get players => _players;
  set players(List<int> value) => _players = value;

  int get minigameId => _minigameId;
  set minigameId(int value) => _minigameId = value;

  factory UniqueGame.fromJson(Map<String, dynamic> json) {
    return UniqueGame(
      json['id'] ?? 0,
      json['highestRound'] ?? 0,
      json['userId'] ?? 0,
      json['isFinished'] ?? false,
      DateTime.parse(json['startTime']),
      (json['players'] as List)?.map((item) => item as int)?.toList() ?? [],
      json['uniqueGameId'] ?? 0,
    );
  }
}
